driver/src/main/java/org/kaazing/k3po/driver/internal/resolver/LocationResolver.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2014, Kaazing Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kaazing.k3po.driver.internal.resolver;
+
+import java.net.URI;
+
+import javax.el.ValueExpression;
+
+import org.kaazing.k3po.driver.internal.behavior.visitor.GenerateConfigurationVisitor;
+import org.kaazing.k3po.lang.internal.ast.value.AstLocation;
+import org.kaazing.k3po.lang.internal.ast.value.AstLocationExpression;
+import org.kaazing.k3po.lang.internal.ast.value.AstLocationLiteral;
+import org.kaazing.k3po.lang.internal.el.ExpressionContext;
+
+/**
+ * The class is used to defer the evaluation of location such as
+ * accept/connect uri in {@link GenerateConfigurationVisitor}. In scenarios when
+ * accept/connect takes expression value which only gets resolved during the
+ * script execution, it is necessary to defer the resolution of accept/connect
+ * uri.
+ *
+ */
+public class LocationResolver {
+
+    private static final LocationVisitorImpl VISITOR = new LocationVisitorImpl();
+
+    private final AstLocation location;
+    private final ExpressionContext environment;
+
+    private URI evaluatedURI;
+
+    public LocationResolver(AstLocation location, ExpressionContext environment) {
+        this.location = location;
+        this.environment = environment;
+    }
+
+    public URI resolve() throws Exception {
+        if (evaluatedURI == null) {
+            evaluatedURI = location.accept(VISITOR, environment);
+        }
+        return evaluatedURI;
+    }
+
+    private static class LocationVisitorImpl implements AstLocation.Visitor<URI, ExpressionContext> {
+
+        @Override
+        public URI visit(AstLocationLiteral value, ExpressionContext parameter) throws Exception {
+            return value.getValue();
+        }
+
+        @Override
+        public URI visit(AstLocationExpression value, ExpressionContext environment) throws Exception {
+            Object uriObj;
+
+            synchronized (environment) {
```

Comment:
I believe this is synchronized because of https://github.com/k3po/k3po/issues/147.  Please add TODO to fix it. I believe all other usages have the same todo